File: src/main/java/pl/reaper/container/data/BankCredit.java
package pl.reaper.container.data;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;
import javax.persistence.Basic;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.xml.bind.annotation.XmlRootElement;

@Entity
@Table(name = "bank_credit")
@XmlRootElement
public class BankCredit implements Serializable {

    private static final long serialVersionUID = 1L;
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Basic(optional = false)
    @Column(name = "id")
    private Integer id;
    @JoinColumn(name = "community_id", referencedColumnName = "id")
    @ManyToOne(optional = false)
    private Community community;
    @JoinColumn(name = "contractor_id", referencedColumnName = "id")
    @ManyToOne(optional = false)
    private Contractor contractor;
    @Column(name = "created_at")
    @Temporal(TemporalType.DATE)
    private Date createdAt;
    @Column(name = "payed")
    private boolean payed;
    @Column(name = "amount")
    private double amount;
    @OneToMany(cascade = CascadeType.PERSIST, mappedBy = "bankCredit")
    private Collection<BankCreditPayment> payments = new ArrayList<>();

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public Community getCommunity() {
        return community;
    }

    public void setCommunity(Community community) {
        this.community = community;
    }

    public Contractor getContractor() {
        return contractor;
    }

    public void setContractor(Contractor contractor) {
        this.contractor = contractor;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public boolean isPayed() {
        return payed;
    }

    public void setPayed(boolean payed) {
        this.payed = payed;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public Collection<BankCreditPayment> getPayments() {
        return payments;
    }

    public void setPayments(Collection<BankCreditPayment> payments) {
        this.payments = payments;
    }

}

File: src/main/resources/scripts/sources/entities/BankCredit.py
from base.Container import Container
from pl.reaper.container.data import BankCredit
from pl.reaper.container.data import BankCreditPayment

class BankCreditManager(Container):

    def create(self):
        credit = BankCredit()
        credit.setCreatedAt(self.parseDate(self._svars.get("createdAt")))
        credit.setAmount(float(self._svars.get("amount")))
        credit.setCommunity(self.findById("Community", self._svars.get("communityId")))
        credit.setContractor(self.findById("Contractor", self._svars.get("contractorId")))
        self.updatePayments(credit)
        self.saveEntity(credit)
        
    def update(self):
        credit = self.findById("BankCredit", self._svars.get('id'))
        self.updatePayments(credit)
        self.saveEntity(credit)

    def remove(self):
        credit = self.findById("BankCredit", self._svars.get('id'))
        if credit.getPayments().size() == 0:
            self._entityManager.remove(credit)
        
    def markAsPayed(self):
        credit = self.findById("BankCredit", self._svars.get('id'))
        credit.setPayed(True)
        self.saveEntity(credit)
        
    def updatePayments(self, credit):
        for i in range(int(self._svars.get('paymentsCount'))): 
            id = int(self._svars.get(str(i) + '_payments_' + 'paymentId'))
            remove = self._svars.get(str(i) + '_payments_' + 'remove') == 'true'
            payment = self.findOrCreatePayment(credit, id)
            payment.setCreatedAt(self.parseDate(self._svars.get(str(i) + '_payments_' + 'createdAt')))
            payment.setAmount(float(self._svars.get(str(i) + '_payments_' + 'amount')))
            if remove:
                self.removePayment(credit, payment)
            
    def findOrCreatePayment(self, credit, id):
        if id == 0:
            payment = BankCreditPayment()
            payment.setBankCredit(credit)
            credit.getPayments().add(payment)
            self._entityManager.persist(payment)
        else:
            payment = self.findById("BankCreditPayment", id);
        return payment
    
    def removePayment(self, credit, payment):
        credit.getPayments().remove(payment)
        self._entityManager.remove(payment)